ironfish-zkp/src/circuits/output.rs

```diff
@@ -0,0 +1,263 @@
+            let esk = boolean::field_into_boolean_vec_le(cs.namespace(|| "esk"), self.esk)?;
+
+            // Create the ephemeral public key from g_d.
+            let epk = g_d.mul(cs.namespace(|| "epk computation"), &esk)?;
+
+            // Expose epk publicly.
+            epk.inputize(cs.namespace(|| "epk"))?;
+        }
+
+        // Now let's deal with pk_d. We don't do any checks and
+        // essentially allow the prover to witness any 256 bits
+        // they would like.
+        {
+            // Just grab pk_d from the witness
+            let pk_d = self
+                .payment_address
+                .as_ref()
+                .map(|e| jubjub::ExtendedPoint::from(*e.pk_d()).to_affine());
+
+            // Witness the v-coordinate, encoded as little
+            // endian bits (to match the representation)
+            let v_contents = boolean::field_into_boolean_vec_le(
+                cs.namespace(|| "pk_d bits of v"),
+                pk_d.map(|e| e.get_v()),
+            )?;
+
+            // Witness the sign bit
+            let sign_bit = boolean::Boolean::from(boolean::AllocatedBit::alloc(
+                cs.namespace(|| "pk_d bit of u"),
+                pk_d.map(|e| e.get_u().is_odd().into()),
+            )?);
+
+            // Extend the note with pk_d representation
+            note_contents.extend(v_contents);
+            note_contents.push(sign_bit);
+        }
+
+        assert_eq!(
+            note_contents.len(),
+            64 + // value
+            256 + // g_d
+            256 // pk_d
+        );
+
+        // Compute the hash of the note contents
+        let mut cm = pedersen_hash::pedersen_hash(
+            cs.namespace(|| "note content hash"),
+            pedersen_hash::Personalization::NoteCommitment,
+            &note_contents,
+        )?;
+
+        {
+            // Booleanize the randomness
+            let rcm = boolean::field_into_boolean_vec_le(
+                cs.namespace(|| "rcm"),
+                self.commitment_randomness,
+            )?;
+
+            // Compute the note commitment randomness in the exponent
+            let rcm = ecc::fixed_base_multiplication(
+                cs.namespace(|| "computation of commitment randomness"),
+                &NOTE_COMMITMENT_RANDOMNESS_GENERATOR,
+                &rcm,
+            )?;
+
+            // Randomize our note commitment
+            cm = cm.add(cs.namespace(|| "randomization of note commitment"), &rcm)?;
+        }
+
+        // Only the u-coordinate of the output is revealed,
+        // since we know it is prime order, and we know that
+        // the u-coordinate is an injective encoding for
+        // elements in the prime-order subgroup.
+        cm.get_u().inputize(cs.namespace(|| "commitment"))?;
+
+        Ok(())
+    }
+}
+
+#[test]
+fn test_output_circuit_with_bls12_381() {
+    use bellman::gadgets::test::*;
+    use ff::Field;
+    use group::Group;
+    use rand_core::{RngCore, SeedableRng};
```

Comment:
```suggestion
    use rand::{RngCore, SeedableRng};
```
Then we can remove the `rand_core` dependency from the `Cargo.toml`